### ISSUE TITLE
Support JSX syntax inside the #call method

### ISSUE THREAD

**Issue by serhiijun** [ViewComponent/view_component#1756]
## Feature request
JSX syntax inside `#call` method for writing HTML.

### Motivation
Currently, the process of incorporating JSX syntax within a ViewComponent involves installing the "rux-rails" gem and renaming the main component file from "my_component.rb" to "my_component.rux". This approach, while functional, introduces an additional step and adds complexity to the development workflow.

I would like to propose the inclusion of native JSX syntax support within ViewComponent itself. This would allow developers to write JSX-styled HTML code directly inside the #call method of a ViewComponent, eliminating the need for external transpilation steps and file renaming.

### Example
```ruby
class Ui::Entries::Home::HomeV1Component::HomeV1Component < ViewComponent::Base
  def initailize()
  end
  
  def call
    <div class={css_class} data-controller={css_class}>
      <Ui::Mix::BarComponent::BarComponent first-name="Bla" last-name="Bla" />
    </div>
  end
  
  def css_class
    @css_class ||= "ui--entries--home--home-v1-component"
  end
end
```

**Comment by Spone** [comment]
Hi, thanks for opening an issue for this! Have you tried the [inline template](https://viewcomponent.org/guide/templates.html#inline) feature? I think it would be more consistent to support JSX with this syntax, rather than reusing `#call`. What do you think?

**Comment by serhiijun** [comment]
@Spone hey thx for superfast reply! Yeah, i already using ViewComponent 3.0 and it may be really great way to use `::InlineTemplate` for support JSX syntax for sure!

**Comment by reeganviljoen** [comment]
@serhiijun have yiu seen rux https://github.com/camertron/rux-rails

**Comment by serhiijun** [comment]
@reeganviljoen sure that's why i propose integrate that as a native support

**Comment by reeganviljoen** [comment]
@serhiijun i have added an issue to gage the interest from the rux people to add their code to view_component

**Comment by reeganviljoen** [comment]
@serhiijun one thing to note is their will be performance optimization to get it working fast

**Comment by serhiijun** [comment]
@reeganviljoen thank you for that, i hope many people will be interested for this feature

**Comment by camertron** [comment]
Hey @reeganviljoen and @serhiijun, thank you for both your interest in view_component and also rux! I happen to be a maintainer of both projects, so this conversation is right up my alley.

> I would like to propose the inclusion of native JSX syntax support within ViewComponent itself.

That would be amazing, but there are some pretty serious technical hurdles we would need to overcome first:

1. Rux (or more specifically rux-rails) uses a series of very tricky monkeypatches to enable the smooth developer experience we have today. The Ruby interpreter does not understand rux code, so it must be transpiled into ruby first. The resulting ruby code is then loaded into the virtual machine. To make that process as frictionless as possible and to prevent developers from having to run the `ruxc` tool constantly, the aforementioned monkeypatches transpile automatically whenever a .rux file is `require`d. There are also monkeypatches in place to ensure .rux files can be reloaded after they change. We would need a much more robust mechanism that does not rely on monkeypatches for me to feel comfortable merging rux into view_component, or for view_component to support syntaxes other than ruby. JavaScript can get away with this much more easily because of the ubiquity of bundlers, but I would bet most Rubyists don't have much of a taste for it.
1. Speaking of the JavaScript ecosystem, another big problem is sourcemaps. Rux stack traces and error messages show transpiled ruby code instead of rux code because again, the ruby interpreter cannot actually _run_ rux code. JavaScript handles this problem by emitting sourcemaps, which map tokens in the transpiled code to tokens in the source code. Browsers have baked in native support for sourcemaps, but the ruby interpreter has not. This means stack traces have to be manually processed before they are shown to users. I'm not even sure what can be done about error messages. Moreover, ruby stack traces do not include column information, meaning the exact location of the error isn't knowable. There is a proposal to add column information in the next version of ruby, but we're not there yet. And when are stack traces processed? Processing all of them is probably too expensive, so you'd have to do it on a case-by-case basis.

Anyway, these are problems I think a lot about. Of course I like rux and want it to succeed, but it feels like we need to solve these problems if rux is going to join ranks with a framework like view_component.

> Have you tried the [inline template](https://viewcomponent.org/guide/templates.html#inline) feature?

It might be a really interesting compromise to use rux in inline templates. For example, what if this were possible?

```ruby
class Ui::Entries::Home::HomeV1Component::HomeV1Component < ViewComponent::Base
  def initailize()
  end
  
  rux_template <<~RUX
    <div class={css_class} data-controller={css_class}>
      <Ui::Mix::BarComponent::BarComponent first-name="Bla" last-name="Bla" />
    </div>
  RUX
  
  def css_class
    @css_class ||= "ui--entries--home--home-v1-component"
  end
end
```

We could probably get syntax highlighting to work in most editors because of the heredoc delimiters, eg. `RUX` in the example above (a delimiter of `RUBY` causes the heredoc to be syntax highlighted as ruby code in VSCode).

**Comment by reeganviljoen** [comment]
I think it would be amazing to get RUX to work in line, but I think getting it to be natively supported in view component as it is would be the best long-term goal, I am interested if a rewrite to make it work more slim under the hood would get us there

**Comment by camertron** [comment]
Right, in-line support would be very cool, but providing `rux_template` doesn't require any of the monkeypatching and gets us most of the way there. I should note it's possible today to create a regular 'ol view file ending in `.html.ruxt` and put rux code into it, which uses Rails' existing template rendering mechanisms. In other words there are multiple options available if you want to use rux without monkeypatches.

**Comment by reeganviljoen** [comment]
@camertron I agree I didn't know that the monkey patches weren't needed for the inline rendering or when the html prefix is used. Is their any reason RUX templates use the `component.rux` form instead of `component.html.rux` form

**Comment by camertron** [comment]
Hey @reeganviljoen, good question. There are two ways to use rux today: inside a Ruby file, and inside a Rails template file. 

Rails template files (i.e. .html.ruxt files) contain only HTML tags (i.e. not component classes). They are processed by Rails' template rendering pipeline, which allows developers to register custom template handlers that are invoked whenever a template file with a particular file extension is requested by the application. The rux-rails gem contains one of these template handlers (i.e. [here](https://github.com/camertron/rux-rails/blob/64d7e4e41bf1e8eb577bf588bbfdf20765dd29c8/lib/rux-rails/template_handler.rb)), which transpiles rux to Ruby. Rails then evaluates this Ruby code whenever the template is rendered.

Ruby files by contrast are loaded into the Ruby interpreter via `Kernel.require` and `Kernel.load`. There is no built-in template handling step as there is with Rails template files. Rux is designed to work with view components, so these Ruby files invariably contain component class definitions. The monkeypatches I referred to earlier hijack the normal behavior of `Kernel.require`, causing it to first transpile the file, _then_ load it into the interpreter.

The main distinction is that there is a built-in mechanism for processing (transpiling) Rails template files, but no such mechanism for Ruby files. This is why monkeypatches are necessary for .rux files, but not for .html.ruxt files.

In our office hours meeting today I chatted with a few of the other view_component maintainers who mentioned that the `rux_template` method should work right now because those method are dynamically generated for each type of template handler. This probably means the right method is actually `ruxt_template`, but I'd have to experiment a bit to verify that.

These *_template methods make it possible to both define the component's template directly inside the component class itself while taking advantage of the built-in Rails template handling mechanism I mentioned earlier. It's _almost_ rux but with none of the monkeypatching.
